cou/steps/plan.py

```diff
@@ -16,25 +16,94 @@
 """Upgrade planning utilities."""
 
 import logging
+from collections import defaultdict
+from typing import List
 
 from cou.steps import UpgradeStep
 from cou.steps.analyze import Analysis
+from cou.steps.application.app import StandardApplication
 from cou.steps.backup import backup
+from cou.utils.openstack import (
+    SPECIAL_CHARMS,
+    UPGRADE_ORDER,
+    CompareOpenStack,
+    determine_next_openstack_release,
+)
 
 logger = logging.getLogger(__name__)
 
 
-async def generate_plan(args: Analysis) -> UpgradeStep:
+async def generate_plan(analysis_result: Analysis) -> UpgradeStep:
     """Generate plan for upgrade.
 
     :param args: Analysis result.
     :type args: Analysis
     :return: Plan with all upgrade steps necessary based on the Analysis.
     :rtype: UpgradeStep
     """
-    logger.info(args)  # for placeholder
+    apps_to_upgrade = determine_apps_to_upgrade(analysis_result)
     plan = UpgradeStep(description="Top level plan", parallel=False, function=None)
     plan.add_step(
         UpgradeStep(description="backup mysql databases", parallel=False, function=backup)
     )
+    for app in apps_to_upgrade:
+        plan.add_step(app.generate_upgrade_plan())
     return plan
+
+
+def determine_apps_to_upgrade(analysis_result: Analysis) -> List[StandardApplication]:
+    """Determine applications to upgrade.
+
+    This function find the oldest OpenStack version in the deployment and
+    select the applications to upgrade for the next version (N + 1).
+
+    :param analysis_result: Analysis result containing all applications in the model.
+    :type analysis_result: Analysis
+    :return: Tuple containing the current and next OpenStack release and a  list
+        of applications to be upgraded.
+    :rtype: Tuple[str, str, List[Application]]
+    """
+    # E.g: {"ussuri": {"keystone"}, "victoria": {"cinder"}}
+    os_versions: defaultdict[str, set] = defaultdict(set)
+
+    for app in analysis_result.apps:
+        # don't consider charms that can have multiple OpenStack releases to determine the oldest
```

Comment:
is this necessary? You're taking the oldest available version in your list so applications that are compatible with multiple o7k codenames should not skew the result

Reply:
You are right. We can remove this logic because it won't affect if we always get the latest compatible version.

Thanks